components/x-gift-article/src/GiftArticle.jsx

```diff
@@ -0,0 +1,117 @@
+import { h } from '@financial-times/x-engine';
+import { withActions } from '@financial-times/x-interaction';
+
+import Loading from './Loading';
+import Form from './Form';
+
+import api from './lib/api';
+import { copyToClipboard } from './lib/share-link-actions';
+import tracking from './lib/tracking';
+
+let hasAttemptedToShortenedNonGiftUrl = false;
+let hasAttemptedToGetAllowance = false;
+let propsComposer;
+
+const withGiftFormActions = withActions(({ articleId, articleUrl, sessionId }) => ({
+	showGiftUrlSection() {
+		return propsComposer.showGiftUrlSection();
+	},
+
+	async showNonGiftUrlSection() {
+		if (propsComposer.isNonGiftUrlShortened) {
+			return propsComposer.showNonGiftUrlSection();
+		} else {
+			const { url, isShortened } = await api.getShorterUrl(articleUrl);
+			if (isShortened) {
+				propsComposer.setShortenedNonGiftUrl(url);
+			}
+			return propsComposer.showNonGiftUrlSection();
+		}
+	},
+
+	async createGiftUrl() {
+		const { redemptionUrl, redemptionLimit } = await api.getGiftUrl(articleId, sessionId);
+
+		if (redemptionUrl) {
+			const { url, isShortened } = await api.getShorterUrl(redemptionUrl);
+			tracking.createGiftLink(url, redemptionUrl);
+			return propsComposer.setGiftUrl(url, redemptionLimit, isShortened);
+		} else {
+			// TODO do something
+		}
+	},
+
+	async getAllowance() {
+		const { giftCredits, monthlyAllowance } = await api.getGiftArticleAllowance();
+
+		// avoid to use giftCredits >= 0 because it returns true when null and ""
+		if (giftCredits > 0 || giftCredits === 0) {
+			return propsComposer.setAllowance(giftCredits, monthlyAllowance);
+		} else {
+			// TODO do something
+		}
+	},
+
+	async getShorterNonGiftUrl() {
+		const { url, isShortened } = await api.getShorterUrl(articleUrl);
+
+		if (isShortened) {
+			return propsComposer.setShortenedNonGiftUrl(url);
+		}
+	},
+
+	copyGiftUrl() {
+		const giftUrl = propsComposer.urls.gift;
+		copyToClipboard(giftUrl);
+		tracking.copyLink('giftLink', giftUrl);
+
+		return propsComposer.showCopyConfirmation();
+	},
+
+	copyNonGiftUrl() {
+		const nonGiftUrl = propsComposer.urls.nonGift;
+		copyToClipboard(nonGiftUrl);
+		tracking.copyLink('nonGiftLink', nonGiftUrl);
+
+		return propsComposer.showCopyConfirmation();
+	},
+
+	emailGiftUrl() {
+		tracking.emailLink('giftLink', propsComposer.urls.gift);
+	},
+
+	emailNonGiftUrl() {
+		tracking.emailLink('nonGiftLink', propsComposer.urls.nonGift);
+	},
+
+	hideCopyConfirmation() {
+		return propsComposer.hideCopyConfirmation();
+	},
+
+	shareByNativeShare() {
+		// TODO display native share
```

Comment:
👍  this can just be an event we emit, something like:

```js
document.body.dispatchEvent(new CustomEvent('xdash.giftArticle.share', { detail: { content: { uuid: 'abc-123', url: 'giftArticleUrl' }}, bubbles: true}));
```

(can fix this properly when we integrate though)

Reply:
I will add this when we integrate to app. Thank you!

Reply:
rather than dispatching an event the app should use `x-interaction` to replace this action